versions_tests/tests/test_models.py

```diff
@@ -1939,3 +1971,186 @@ def test_filter_on_fk_relation(self):
         team.clone()
         l2 = len(Player.objects.as_of(t1).filter(team__name='team'))
         self.assertEqual(l1, l2)
+
+class SpecifiedUUIDTest(TestCase):
+
+    @staticmethod
+    def uuid4():
+        return six.text_type(str(uuid.uuid4()))
+
+    def test_create_with_uuid(self):
+        p_id = self.uuid4()
+        p = Person.objects.create(id=p_id, name="Alice")
+        self.assertEqual(p_id, p.id)
+        self.assertEqual(p_id, p.identity)
+
+        p_id = six.text_type(str(uuid.uuid5(uuid.NAMESPACE_OID, 'bar')))
+        with self.assertRaises(ValueError):
+            Person.objects.create(id=p_id, name="Alexis")
+
+    def test_create_with_forced_identity(self):
+
+        # This test does some artificial manipulation of versioned objects, do not use it as an example
+        # for real-life usage!
+
+        p = Person.objects.create(name="Abela")
+
+        # Postgresql will provide protection here, since util.postgresql.create_current_version_unique_identity_indexes
+        # has been invoked in the post migration handler.
+        if connection.vendor == 'postgresql' and get_version() >= '1.7':
+            with self.assertRaises(IntegrityError):
+                with transaction.atomic():
+                    Person.objects.create(forced_identity=p.identity, name="Alexis")
+
+        p.delete()
+        sleep(0.1)  # The start date of p2 does not necessarily have to equal the end date of p.
+
+        p2 = Person.objects.create(forced_identity=p.identity, name="Alexis")
+        p2.version_birth_date = p.version_birth_date
+        p2.save()
+        self.assertEqual(p.identity, p2.identity)
+        self.assertNotEqual(p2.id, p2.identity)
+
+        # Thanks to that artificial manipulation, p is now the previous version of p2:
+        self.assertEqual(p.name, Person.objects.previous_version(p2).name)
+
+
+class VersionRestoreTest(TestCase):
+
+    def setup_common(self):
+        sf = City.objects.create(name="San Francisco")
+        forty_niners = Team.objects.create(name='49ers', city=sf)
+        player1 = Player.objects.create(name="Montana", team=forty_niners)
+        best_quarterback = Award.objects.create(name="Best Quarterback")
+        best_attitude = Award.objects.create(name="Best Attitude")
+        player1.awards.add(best_quarterback, best_attitude)
+
+        self.player1 = player1
+        self.awards = {
+            'best_quarterback': best_quarterback,
+            'best_attitude': best_attitude,
+        }
+        self.forty_niners = forty_niners
+
+    def testRestoreLatestVersion(self):
```

Comment:
Could you please use underscores for method names (and CamelCase only for class names)? Thanks ;)